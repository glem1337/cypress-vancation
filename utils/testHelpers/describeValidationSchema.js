export const validationSchema = (wrapper) => {
  if ('validationSchema' in wrapper.props()) {
    return wrapper.props().validationSchema();
  }

  return validationSchema(wrapper.dive());
};

const describeValidationSchema = wrapper => (
  validationSchema(wrapper).describe()
);

export default describeValidationSchema;
