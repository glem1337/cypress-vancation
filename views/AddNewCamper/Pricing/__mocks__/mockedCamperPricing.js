export default {
  costomizialeNightCost: false,
  weeklyDiscount: false,
  monthlyDiscount: false,
  weeklyDiscountPercent: 15,
  monthlyDiscountPercent: 33,
  minimalNightStay: '2',
  costPerNight: 280,
  weekNightPrice: {
    monday_price: '250',
    tuesday_price: '250',
    wednesday_price: '250',
    thursday_price: '250',
    friday_price: '320',
    saturday_price: '320',
    sunday_price: '250',
  },
};
