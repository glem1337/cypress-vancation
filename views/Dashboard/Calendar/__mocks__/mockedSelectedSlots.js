export default {
  start: new Date(),
  end: new Date(),
  slots: [new Date()],
  uuid: 'uuid',
};
