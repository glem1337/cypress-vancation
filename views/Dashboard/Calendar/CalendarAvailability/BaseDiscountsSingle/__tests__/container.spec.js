import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import * as R from 'ramda';

import diveTo from 'utils/testHelpers/diveToEnzyme';
import describeValidationSchema from 'utils/testHelpers/describeValidationSchema';
import mockedCamperPricing from 'views/AddNewCamper/Pricing/__mocks__/mockedCamperPricing';
import mockedSelectedSlots from 'views/Dashboard/Calendar/__mocks__/mockedSelectedSlots';

import BaseDiscountSingle, { BaseDiscountSingleContainer } from '../container';

jest.mock('state/concepts/camper/selectors', () => ({
  camperPricingSelector: jest.fn(() => mockedCamperPricing),
}));

jest.mock('state/concepts/calendar/selectors', () => ({
  selectedSlotsSelector: jest.fn(() => mockedSelectedSlots),
  customBaseDiscountsForSingleDate: jest.fn(() => ({
    weeklyDiscountPercent: null,
    monthlyDiscountPercent: null,
  })),
}));

jest.mock('state/data/selectors', () => ({
  loadingSelector: jest.fn(() => true),
}));

jest.mock('formik', () => ({
  withFormik: (config) => (Component) => props => {
    const extended = {
      ...props,
      ...config,
      validationSchema: () => config.validationSchema,
    };

    return <Component {...extended} />;
  },
}));

jest.mock('utils/calendar/withPopovers', () => jest.fn((Component) => (props) => (
  <Component
    {...props}
    isOpened={false}
    isBackPopoverVisible={false}
    isDiscardPopoverVisible={false}
    isSavePopoverVisible={false}
    checkOpenedState={jest.fn()}
    showPopover={jest.fn()}
    showPopoverOrBack={jest.fn()}
    closePopover={jest.fn()}
    closeAllPopovers={jest.fn()}
    closePopoversAndBack={jest.fn()}
    closePopoversAndSave={jest.fn()}
    toggleOpenedState={jest.fn()}
  />
)));

describe('BaseDiscountSingle container tests', () => {
  const store = configureStore()({});
  store.dispatch = jest.fn();

  const props = {
    store,
    values: { isOpened: true },
    setFieldValue: jest.fn(),
    isLoading: false,
    isSubmitting: false,
    submitForm: jest.fn(),
    isValid: true,

  };

  let wrapper = null;
  let container = null;
  let instance = null;

  beforeEach(() => {
    wrapper = shallow(<BaseDiscountSingle {...props} />);
    container = diveTo(wrapper, BaseDiscountSingleContainer);
    instance = container.instance();

    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    expect(container).toMatchSnapshot();
  });

  it('has validation schema', () => {
    expect(describeValidationSchema(wrapper)).toMatchSnapshot();
  });

  it('checks mapPropsToValues configs', () => {
    expect(container.props().mapPropsToValues((container.props()))).toMatchSnapshot();
  });

  it('checks `componentDidUpdate` instance method', () => {
    instance.componentDidUpdate(props);

    expect(instance.props.checkOpenedState).toHaveBeenCalled();
  });

  it('tests `weekCost` instance getter', () => {
    const {
      values: { costPerNight, customizableNightCost, weekNightPrice },
    } = props;

    const expected = customizableNightCost
      ? R.compose(R.sum, R.values)(weekNightPrice)
      : costPerNight * 7;

    expect(instance.weekCost).toEqual(expected);
  });

  it('tests `weeklyDiscountPercentPrice` instance getter', () => {
    const {
      values: { weeklyDiscountPercent },
    } = props;

    const expected = Math.round(
      (instance.weekCost / 100) * weeklyDiscountPercent,
    );

    expect(instance.weeklyDiscountPercentPrice).toEqual(expected);
  });

  it('tests `monthlyDiscountPercentPrice` instance getter', () => {
    const {
      values: { monthlyDiscountPercent },
    } = props;

    const monthCost = instance.weekCost * 4;

    const expected = Math.round((monthCost / 100) * monthlyDiscountPercent);

    expect(instance.monthlyDiscountPercentPrice).toEqual(expected);
  });
});
