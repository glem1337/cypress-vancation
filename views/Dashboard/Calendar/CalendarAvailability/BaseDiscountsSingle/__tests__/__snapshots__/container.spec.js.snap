// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseDiscountSingle container tests checks mapPropsToValues configs 1`] = `
Object {
  "costPerNight": 280,
  "costomizialeNightCost": false,
  "monthlyDiscount": false,
  "monthlyDiscountPercent": 33,
  "weekNightPrice": Object {
    "friday_price": "320",
    "monday_price": "250",
    "saturday_price": "320",
    "sunday_price": "250",
    "thursday_price": "250",
    "tuesday_price": "250",
    "wednesday_price": "250",
  },
  "weeklyDiscount": false,
  "weeklyDiscountPercent": 15,
}
`;

exports[`BaseDiscountSingle container tests has validation schema 1`] = `
Object {
  "fields": Object {
    "monthlyDiscountPercent": Object {
      "label": undefined,
      "meta": undefined,
      "notOneOf": Array [],
      "oneOf": Array [],
      "tests": Array [
        Object {
          "name": "required",
          "params": undefined,
        },
        Object {
          "name": "min",
          "params": Object {
            "min": 1,
          },
        },
        Object {
          "name": "max",
          "params": Object {
            "max": 99,
          },
        },
      ],
      "type": "number",
    },
    "weeklyDiscountPercent": Object {
      "label": undefined,
      "meta": undefined,
      "notOneOf": Array [],
      "oneOf": Array [],
      "tests": Array [
        Object {
          "name": "required",
          "params": undefined,
        },
        Object {
          "name": "min",
          "params": Object {
            "min": 1,
          },
        },
        Object {
          "name": "max",
          "params": Object {
            "max": 99,
          },
        },
      ],
      "type": "number",
    },
  },
  "label": undefined,
  "meta": undefined,
  "notOneOf": Array [],
  "oneOf": Array [],
  "tests": Array [],
  "type": "object",
}
`;

exports[`BaseDiscountSingle container tests matches snapshot 1`] = `
<BaseDiscountSingleComponent
  checkOpenedState={[MockFunction]}
  closeAllPopovers={[MockFunction]}
  closePopover={[MockFunction]}
  closePopoversAndBack={[MockFunction]}
  closePopoversAndSave={[MockFunction]}
  customBaseDiscounts={
    Object {
      "monthlyDiscountPercent": null,
      "weeklyDiscountPercent": null,
    }
  }
  enableReinitialize={true}
  handleSubmit={[Function]}
  isBackPopoverVisible={false}
  isDiscardPopoverVisible={false}
  isLoading={true}
  isOpened={false}
  isRange={false}
  isSavePopoverVisible={false}
  isSubmitting={false}
  isValid={true}
  mapPropsToValues={[Function]}
  monthlyDiscountPercentPrice={NaN}
  onSubmit={[Function]}
  pricingInfo={
    Object {
      "costPerNight": 280,
      "costomizialeNightCost": false,
      "minimalNightStay": "2",
      "monthlyDiscount": false,
      "monthlyDiscountPercent": 33,
      "weekNightPrice": Object {
        "friday_price": "320",
        "monday_price": "250",
        "saturday_price": "320",
        "sunday_price": "250",
        "thursday_price": "250",
        "tuesday_price": "250",
        "wednesday_price": "250",
      },
      "weeklyDiscount": false,
      "weeklyDiscountPercent": 15,
    }
  }
  selectedSlots={
    Object {
      "end": 2000-02-20T00:00:00.000Z,
      "slots": Array [
        2000-02-20T00:00:00.000Z,
      ],
      "start": 2000-02-20T00:00:00.000Z,
      "uuid": "uuid",
    }
  }
  setFieldValue={[MockFunction]}
  showPopover={[MockFunction]}
  showPopoverOrBack={[MockFunction]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  submitForm={[MockFunction]}
  toggleOpenedState={[MockFunction]}
  validationSchema={[Function]}
  values={
    Object {
      "isOpened": true,
    }
  }
  weeklyDiscountPercentPrice={NaN}
/>
`;
